File: Shashank Shukla 18-CSE/Day-7.java
/*
Time Complexity: O(N)
Space Complexity: O(N)
*/

class Solution {
    public boolean find132pattern(int[] nums) {
        Stack<Integer> st = new Stack<>();
        int n = nums.length;
        int last = Integer.MIN_VALUE;
        for(int i= n-1; i>=0; i--){
            if(last>nums[i]){
                return true;
            }
            while(!st.isEmpty() && st.peek()<nums[i])
                last = st.pop();
            st.push(nums[i]);
        }
        return false;
    }
}

File: Aman Dwivedi 18-ECE/Day 7.cpp
https://leetcode.com/problems/132-pattern/
Time: O(N)
Space: O(N)

class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        
        int third = INT_MIN;
        int n=nums.size();
        stack<int> st;
        for(int i=n-1;i>=0;i--)
        {
            if(nums[i]<third)
                return true;
            while(!st.empty()&&nums[i]>st.top())
            {
                third=st.top();
                st.pop();
            }
            st.push(nums[i]);
        }
        return false;
    }
    
};
